src/json.ts

```diff
@@ -4,9 +4,110 @@ export type JSONValue =
   | boolean
   | number
   | Array<JSONValue>
-  // undefined are filtered out in JSON.stringify so OK to allow.
-  | {[key: string]: JSONValue | undefined};
+  | JSONObject;
+
+type JSONObject = {[key: string]: JSONValue};
 
 export type ToJSON = {
-  toJSON: () => JSONValue;
+  toJSON: (key?: string) => JSONValue;
 };
+
+/**
+ * Like [[JSONValue]] but allows [[ToJSON]] at every recursive value. This
+ * matches the semantics of `JSON.stringify`.
+ */
+export type ToJSONValue =
+  | ToJSON
+  | null
+  | string
+  | boolean
+  | number
+  | Array<ToJSONValue>
+  | ToJSONObject;
+
+type ToJSONObject = {[key: string]: ToJSONValue};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function hasToJSON(v: any): v is ToJSON {
+  return typeof v?.toJSON === 'function';
+}
+
+/**
+ * Does a deep equality check on a JSON value with same semantics as
+ * JSON.stringify (except for the ordering of the keys in an Object).
```

Comment:
I don't understand the parenthetical in this comment. Can you expand it?

Reply:
Yup.

The semantics of JSON.stringify is to keep the order of the keys. `JSON.stringify({a: 1, b: 2})` => `'{"a":1,"b":2}'` but `JSON.stringify({b: 2, a: 1})` => `'{"b":2,"a":1}'`. This equality function treats `{a: 1, b: 2}` and `{b: 2, a: 1}` as equal.

Let me see if I can think of a concise and clear way to express that.